### DISCUSSION TITLE
Creating palette image on the fly for remap behind : palette is not reset each time

### DISCUSSION THREAD

**Discussion by chridubois** [ImageMagick/ImageMagick#6332] [Help] [answered]
### ImageMagick version

7.1.1-9

### Operating system, version and so on

MacOS

### Description

Hi, 

I'm using Rmagick (Ruby on Rails gem to use ImageMagick).
I'm building a tool to remap images with some specific color palette.

As inputs, I have : 
- An image (.png or .jpg)
- A array of color hex codes.

I've built this script to create the palette with color hex codes in order to use it during the remap : 
`colorpallet = MiniMagick::Image.new("palette_tmp.png")`
    `colors.each do |color_item|`
      `color = MiniMagick::Image.new("color_tmp.png")`
      `color.run_command('convert', '-size', '1x1', "xc:#{color_item}", color.path)`
      `colorpallet.run_command('convert', '+append', colorpallet.path, color.path, colorpallet.path)`
    `end`

When I run my script, I get this error : 

> `magick convert +append palette_tmp.png color_tmp.png palette_tmp.png` failed with status: 1 and error:
> convert: unable to open image 'palette_tmp.png': No such file or directory @ error/blob.c/OpenBlob/3569.

Yet, palette_tmp.png is existing...

Then, i use the remap operator like this : 
`remap_sketch.run_command('convert', url, '-dither', 'FloydSteinberg', '-remap', colorpallet.path, remap_sketch.path)`

I have two questions : 

1) Do you know why I have this error about "No such file or directory" ?

2) Is there a way to do a remap without generating a color palette file on the fly ?

Thanks a lot in advance.

Christophe

**Comment by snibgo** [comment]
I don't use or know Rmagick.

I suggest you rearrange commands so the options are in the order they should be processed. Put "`+append`" after reading the two inputs, not before.

I don't know why you get "`No such file or directory`".

You use two convert commands and use a temporary file, when you need only one command and no temporary file.

Your two commands might be:
```
convert -size 1x1 xc:#abc color_tmp.png
convert palette_tmp.png color_tmp.png +append palette_tmp.png
```
Instead, you could do this:
```
convert palette_tmp.png -size 1x1 xc:#abc +append palette_tmp.png
```
Even better, you could build the `convert` command in a loop, and execute it only once. The command could be like this:
```
convert palette_tmp.png -size 1x1 xc:#abc xc:#123 xc:#234 xc:#345 +append palette_tmp.png
```

**Comment by chridubois** [comment]
Hi snibgo,

Thank you very much for your answer.

I've just tried this code : 

`    colorpallet = MiniMagick::Image.new("palette_tmp.png")`
`    colors.each do |color_item|`
`      colorpallet.run_command('convert', colorpallet.path, '-size', '1x1', "xc:#{color_item}", '+append', colorpallet.path)`
`    end`

and I still have the same issue : 

<img width="1047" alt="image" src="https://github.com/ImageMagick/ImageMagick/assets/113714569/5f451522-3d6d-427c-b287-1a27454295a1">

Yet, I confirm that 'palette_tmp.png' is well existing on my folder...

Do you have any other ideas?

I had the idea to build a txt file of color palette instead of an image, but I did not succeed ...

Thanks a lot

**Reply by fmw42** [reply]
On IM 7, use magick, not magick convert and not convert.

**Comment by chridubois** [comment]
Hi fmw42,

You're right, I've not this error anymore ! 

Now, I get an other issue : my 'palette_tmp.png' is never reseted.

Means that, 
`colorpallet = MiniMagick::Image.new("palette_tmp.png")`
looks not erasing the old one.

Result : each time I run the script for different uploaded image, ma color palette is beginning bigger and bigger.

Do we have a way to reset an image file ?

I've tried this way : 

`    File.delete("palette_tmp.png") if File.exist?("palette_tmp.png")`
`    colorpallet = MiniMagick::Image.new("palette_tmp.png")`
`    colors.each do |color_item|`
`      colorpallet.run_command('magick', colorpallet.path, '-size', '1x1', "xc:#{color_item}", '+append', colorpallet.path)`
`    end`

But doing this, I have the same erreur than before 👍 

> magick: unable to open image 'palette_tmp.png': No such file or directory @ error/blob.c/OpenBlob/3569.

Thanks !

**Reply by fmw42** [reply]
Looks to me like the palette is not ready after you create it for use in the subsequent command.  Also MiniMagick seems to not remove the old one and appends to the previous palette. 

I would suggest you consult with the Minimagick creators or forum.  Sorry, I do not use MiniMagick.

**Comment by snibgo** [comment]
Perhaps the problem is that you are mixing two methods of programming. One uses RMagick. The other uses shell commands. Perhaps the file `palette_tmp.png` is opened by Rmagick in an exclusive mode, so when the `magick` command tries to open it, it is locked out.

You can test this by temporarily removing the line:
```
colorpallet = MiniMagick::Image.new("palette_tmp.png")
```
If that is the problem, then you could use the Rmagick operation that appends images.

**Comment by chridubois** [comment]
I think I understand the point yes. 
But if I remove the following line : colorpallet = MiniMagick::Image.new("palette_tmp.png")
I will not be able to use "run_command()"

Do you know how can I run Shell commands with Ruby on Rails ?

Thanks

**Reply by fmw42** [reply]
Sorry, I do not know Ruby onRails or Minimagick.  Can you not just use the shell command directly?

**Reply by chridubois** [reply]
yes, When I use the Shell command I still have this error : 

`convert: unable to open image 'palette_tmp.png': No such file or directory @ error/blob.c/OpenBlob/3569.`
<img width="720" alt="image" src="https://github.com/ImageMagick/ImageMagick/assets/113714569/7afc8ab6-4ba6-482d-8b3a-23021b064c36">

(I've not put existing color for the test)

**Reply by fmw42** [reply]
Don't use convert on IM 7. Use magick.  Please show your command for how you create your palette.png prior to your command above.

Please do not post images of code or your directory. It makes it hard to re-write your code or test it.

On Mac OSX put your hex colors in double quotes.

**Reply by chridubois** [reply]
Sorry, using magick instead of convert, it's working : 

`magick palette_tmp.png -size 1x1 xc:#abc xc:#123 xc:#234 xc:#345 +append palette_tmp.png`

But my palette is still appending new colors (keeping de old ones)

After 3 times running : 
<img width="240" alt="image" src="https://github.com/ImageMagick/ImageMagick/assets/113714569/d49d888a-c560-469a-a812-bb4f88e2161b">

**Reply by fmw42** [reply]
Of course it will.  You append new colors to your existing palette and save with the **same** input name.  If you don't want that, change the name of the output.  Then the output will have new colors, but the input will be unchanged.

**Reply by chridubois** [reply]
But I would like to have only one tmp file which is reseted each time I run the script.
In order to not store a lot of palette, unnecessary.

I thought It was possible to restart from scratch before each script. 
Maybe there is a command for deleting palette file just before running the script ? That's what I've tried upper

I've just tried to delete palette file manually before running the command. And I still get same issue : 

> magick: unable to open image 'palette_tmp.png': No such file or directory @ error/blob.c/OpenBlob/3569.

**Reply by fmw42** [reply]
What are you trying to do?  Please explain.  What is your code for creating your initial palette image and what do you want to do with the palette image once created?

**Reply by fmw42** [reply]
A palette image can be created in-line for -remap without saving it. Is this what you want?  It creates a colortable image and saves to MPR: in-memory format without writing to disk.  Then uses the MPR: image for the -remap.

`magick \( xc:"#hex_color1" xc:"#hex_color2" xc:"#hex_color3" ... xc:"#hex_colorN" +append -write mpr:colortable +delete \) input_image_to_recolor -remap mpr:colortable output_image_after_recolor`

**Reply by chridubois** [reply]
The goal is to remap an uploaded image with a selected color palette.
User's inputs are : 
- An image (jpg or png)
- A list of color hex codes

I have a code to get user's hex codes list.

In order to use the remap operation, I need to build a palette file with those hex codes.

**Here is how I do :** 

I create a palette file with : `colorpallet = MiniMagick::Image.new("palette_tmp.png")`

For each hex_code from the list : 
I create a color file and I append this color to the palet file with : `convert palette_tmp.png -size 1x1 xc:#abc +append palette_tmp.png`

When I have my palette, I open it with : `colorpallet = MiniMagick::Image.open(palette)`
I create a "remap Image" file with : remap_sketch = MiniMagick::Image.new('sketch_tmp.png')
And I remap with this command : 'magick', original_image, '-dither', 'FloydSteinberg', '-remap', colorpallet.path, remap_sketch.path)

**Reply by chridubois** [reply]
> A palette image can be created in-line for -remap without saving it. Is this what you want? It creates a colortable image and saves to MPR: in-memory format without writing to disk. Then uses the MPR: image for the -remap.
> 
> `magick \( xc:"#hex_color1" xc:"#hex_color2" xc:"#hex_color3" ... xc:"#hex_colorN" +append -write mpr:colortable +delete \) input_image_to_recolor -remap mpr:colortable output_image_after_recolor`

Yes it's exactly what I would like to do !

But I need to do this in my Ruby on Rails controller, not in Shell...

To do it in Ruby on Rails, I used to use command like : 

`remap_sketch.run_command('convert', url, '-dither', 'FloydSteinberg', '-remap', colorpallet.path, remap_sketch.path)`

But i don't succeed to make your complex one work with this syntaxe

**Reply by fmw42** [reply]
I doubt Ruby on Rails is sophisticated enough to handle the MPR: processing or perhaps not parentheses processing.  It likely wants a disk image to process. That means you need to create a palette image and save to disk. Then use it in your -remap command.   Somehow you need to close out your process, if you want to re-use the same palette name.

Sorry, I do not know Ruby on Rails. Please consult with the Ruby on Rails forum or developers.

Again do not use convert, use magick on IM 7.

**Reply by chridubois** [reply]
Yes I will do that Thanks !

But do you know why this command is not working on shell : 

`magick palette_tmp.png -size 1x1 xc:#abc xc:#123 xc:#234 xc:#345 +append palette_tmp.png`

We should create a file before ?

**Reply by fmw42** [reply]
In what way is it not working?  Again, you are saving the output image with the same name as the input image. So it will keep appending the colors repeatedly.  I keep asking -- How do you create your original palette image?  Please answer! So that I can understand your full process of what you are trying to do.  How about

**Reply by chridubois** [reply]
I'm creating it with this command : 

`colorpallet = MiniMagick::Image.new("palette_tmp.png")`

**Comment by snibgo** [comment]
> Means that,
colorpallet = MiniMagick::Image.new("palette_tmp.png")
looks not erasing the old one.

No, it doesn't. That command reads a file called `palette_tmp.png`. It doesn't erase the file, and it doesn't create an empty image.

@chridubois: Why are you using `run_command` as well as Rmagick operations such as `Image.new`? I think that is your problem. I suggest you either use only `run_command` for *all* your operations, or use only Rmagick commands such as `ilist.append` and `ilist.remap`. See:

https://rmagick.github.io/ilist.html#append
https://rmagick.github.io/ilist.html#remap

How did you create the file `palette_tmp.png`?

**Reply by chridubois** [reply]
I will check thanks ! 

I've created it only with this command : 

`colorpallet = MiniMagick::Image.new("palette_tmp.png")`

As I understand, to open a file it's an other command : 

`colorpallet = MiniMagick::Image.open("palette_tmp.png")`

**Reply by chridubois** [reply]
Thanks for the documentation.

I tried this code : 

`    remap_image = Magick::Image.new(256, 64)`
`    ilist = Magick::ImageList.new`
`    colors.each do |color_item|`
`      ilist.new_image(1, 1) { |options| options.background_color = "#{color_item}" }`
`    end`
`    ilist.remap(remap_image, dither='RiemersmaDitherMethod')`

But I don't know how to see if the ilist is well built.
And it returns me this error : 

> wrong enumeration type - expected Magick::DitherMethod, got String

for the line : 
`ilist.remap(remap_image, dither='RiemersmaDitherMethod')`

**Comment by snibgo** [comment]
> ilist.remap(remap_image, dither='RiemersmaDitherMethod')

> wrong enumeration type - expected Magick::DitherMethod, got String

It expects a constant, not a string. Why do you have quotation marks?

**Reply by chridubois** [reply]
Hum, I don't really understand.

I've created a env constant : DITHER = RiemersmaDitherMethod

and 

`    ilist.remap(remap_image, DITHER)`

still not work with same error : 

> wrong enumeration type - expected Magick::DitherMethod, got String

**Reply by fmw42** [reply]
Check the documentation for Minimagick. I suspect it wants a numerical value 0,1,2 etc in place of the text name you provided.

**Reply by chridubois** [reply]
Hi fmw42,

I've checked documentation [https://rmagick.github.io/ilist.html#remap] but they only say this : 

"ilist.remap(remap_image=nil, dither=RiemersmaDitherMethod) -> self"

I've tried with 0,1,2 but I have this error : 

> wrong enumeration type - expected Magick::DitherMethod, got Integer

**Comment by snibgo** [comment]
I guess it just needs the quotation makes removed.
```
ilist.remap(remap_image, dither=RiemersmaDitherMethod)
```
But I don't know Ruby.

**Reply by chridubois** [reply]
Hi snibgo,

When I put what you mentionned, i've got this error : 

> uninitialized constant SketchesController::RiemersmaDitherMethod

And when I initialized the constant, i get back to the last error :

> wrong enumeration type - expected Magick::DitherMethod, got String

**Comment by snibgo** [comment]
> uninitialized constant SketchesController::RiemersmaDitherMethod

Note the namespace where it is searching for the constant, `SketchesController`.

But it wants `Magick::DitherMethod`.

So try giving the `Magick` namespace:

```
ilist.remap(remap_image, dither=Magick::RiemersmaDitherMethod)
```

**Reply by chridubois** [reply]
Yes I've already done that and the error was the same than when I put a string with:

wrong enumeration type - expected Magick::DitherMethod, got String

**Reply by fmw42** [reply]
Perhaps a bug in R Magick or Minimagick.  Should you not consult with the developers of those tools or one of their forums?

**Reply by chridubois** [reply]
Finally, it worked ! 
I've been code : 
`ilist.remap(remap_image, Magick::RiemersmaDitherMethod)`
instead of 
`ilist.remap(remap_image, dither=Magick::RiemersmaDitherMethod)`

sorry snibgo, you was right !

My final step is to store this remap image into Cloudinary in order to display it on my front-end. I did not find how to get path of the remap_image.
Compared to my last CLI command with Magick, those commands are not creating any files.

I'm trying with something like : 

`temp = Tempfile.new("image")
remap_image.write("jpeg:"+ temp.path)`

**Comment by chridubois** [comment]
Thanks to you guys, I've succeeded to run far away from the beginning.
But I still not succeed to remap my image perfectly !

Let me give you the example.

First, my code : 
```
image = Magick::Image.read("http://res.cloudinary.com/dzqnghvr3/image/upload/e_vectorize:colors:10:detail:500/v1/development/#{@sketch.photo.key}")
    ilist = Magick::ImageList.new
    
    # Create palette with Magick ilist
    colors.each do |color_item|
      ilist.new_image(100, 100) { |options| options.background_color = "#{color_item}" }
    end

    # Store palette image in a file and in cloud
    palette = Tempfile.new("palette")
    ilist.write("gif:"+ palette.path)
    Cloudinary::Uploader.upload(palette.path, :folder => "development/", :public_id => "p_#{@sketch.photo.key}", :overwrite => true)

    # Remap original image
    image.first.remap(ilist, dither=Magick::RiemersmaDitherMethod)
    
    # Store remaped image in a file and in cloud
    remap_image = Tempfile.new("remap_image")
    image.first.write("png:"+ remap_image.path)
    Cloudinary::Uploader.upload(remap_image.path, :folder => "development/", :public_id => "rc_#{@sketch.photo.key}", :overwrite => true)
```


In this example, i'm trying to remap an image with 4 colours : 
![ulz49deve654omod3r2gp3ieptrg](https://github.com/ImageMagick/ImageMagick/assets/113714569/7c52f9b4-e271-4fa5-b756-000cc56d45f0)

with a palette of two colours (randomly picked from the image above) : 
![palette](https://github.com/ImageMagick/ImageMagick/assets/113714569/8be9d4a7-02dc-4ce0-b29d-5ecb629d444a)

The result of the remaped image is : 
![rc_ulz49deve654omod3r2gp3ieptrg](https://github.com/ImageMagick/ImageMagick/assets/113714569/a074d4c1-a335-441c-9887-0029c8c91c2c)

As I can see, the remap works ONLY with the last color of the palette.
(Means that when my palette contains 1 color, it's working)

Do you know why ?

Edit : I've tried to change Dither method of the remap... But nothing change

**Reply by fmw42** [reply]
Don't use an animated GIF for the palette.  Use a normal gif with appended colors or a png with appended colors. But not an animated GIF with multiple frames.

**Reply by chridubois** [reply]
According to [https://rmagick.github.io/ilist.html#write], a ilist can't be a non-animated file...

I've tried to do : 
`# Store palette image in a file and cloud
    palette = Tempfile.new("palette")
    ilist.write("jpg:"+ palette.path)`

And I get this error : Empty file

Same with .png

I should maybe not use ilist finally...

**Reply by chridubois** [reply]
I've just succeeded to replace animated gif palette with static palette with this code : 

`# Store palette image in a file and cloud
    palette = ilist.append(false)
    palette_file = Tempfile.new("palette")
    palette.write("jpg:"+ palette_file.path)`

It generates me this palette image : 
![palette](https://github.com/ImageMagick/ImageMagick/assets/113714569/11e9e63a-efa5-45b5-995c-1d57b978a557)

But still not remaping... Same output

**Comment by snibgo** [comment]
Show your full code, not just a couple of lines.

**Reply by chridubois** [reply]
Here it is : 

```
def user_palette
    # Get Palette colors
    colors = params[:colors].split(',')

    # Get uploaded image
    image = Magick::Image.read("http://res.cloudinary.com/dzqnghvr3/image/upload/e_vectorize:colors:10:detail:500/v1/development/#{@sketch.photo.key}")
    
    # Create a new ilist
    ilist = Magick::ImageList.new

     # Fill the inlist with colors
    colors.each do |color_item|
      ilist.new_image(100, 100) { |options| options.background_color = "#{color_item}" }
    end

    # Store palette image in a file and cloud
    palette = ilist.append(false)
    palette_file = Tempfile.new("palette")
    palette.write("jpg:"+ palette_file.path)
    Cloudinary::Uploader.upload(palette_file.path, :folder => "development/", :public_id => "p_#{@sketch.photo.key}", :overwrite => true)

    # Remap original image
    image.first.remap(ilist, dither=Magick::FloydSteinbergDitherMethod)

    # Store remap image in a file and cloud
    remap_image = Tempfile.new("remap_image")
    image.first.write("png:"+ remap_image.path)
    Cloudinary::Uploader.upload(remap_image.path, :folder => "development/", :public_id => "rc_#{@sketch.photo.key}", :overwrite => true)
  end
```

**Comment by snibgo** [comment, marked as answer]
You are remapping with `ilist`, which is a list of images. You have appended those images together into `palette`, but you don't use that for remapping.

EDIT: I should also mention that you store the palette image, which has 100x100 pixel squares of colours, in a JPG which is a lossy format. This is probably a bad idea. If your have, say, 5 pixel squares the JPG file may contain 50 or 500 colours.

**Reply by chridubois** [reply]
Hi snibgo,

You was right. Everything is working well now.

Thanks to you and fmw42.

You're the best !
